src/authorize.ts

```diff
@@ -0,0 +1,33 @@
+import * as express from 'express';
+
+export interface IEveryAuthContext {
+  finishedUrl: string;
+  identity: any;
+}
+
+export interface IEveryAuthOptions {
+  finishedUrl: string;
+  mapToUserId: (req: express.Request) => Promise<string>;
+  mapToTenantId?: (req: express.Request) => Promise<string>;
+  onAuthorized?: (req: express.Request, res: express.Response, everyCtrx: IEveryAuthContext) => Promise<void>;
+  onComplete?: (req: express.Request, res: express.Response, everyCtrx: IEveryAuthContext) => Promise<void>;
+}
+
+export const authorize = (serviceId: string, options: IEveryAuthOptions): express.Router => {
+  const router = express.Router();
+
+  router.get('/', (req: express.Request, res: express.Response, next: express.NextFunction) => {
+    // Create the session with the redirectUrl as '/complete';
+    // req.redirect(session.nextUrl);
+  });
+
+  router.get('/complete', (req: express.Request, res: express.Response, next: express.NextFunction) => {
+    // Call options.onAuthorized with the nascient session object
+    // Update the session object if it's changed
+    // session.commit()
+    // Check for error, and propagate to redirect
+    // res.redirect(options.finishedUrl);
+  });
```

Comment:
Everyone _needs_ to supply a finish URL, which makes sense that it's a required property.  As for the sequence changing, this is the current layout":
* The onAuthorized allows them to modify (add tags, etc) prior to "commit".
* Then the commit happens, which is (and should be) opaque and non-optional (baring error) to the developer.
* The onComplete allows them to capture keys (save the install id, for example) after commit.
* Finally, finishedUrl is how the middleware passes execution control back to "the app".

Is there a different sequence that makes sense?

Reply:
The default finish URL could be defined by convention - if they hang the middleware on /slack, it could be /slack/finish. So the common case would look like this:

```javascript
app.use('/slack', everyauth.authorize('slack', {
   mapToUserId: async (req) => 'usr-123
}));

app.get('/slack/complete', (req, res, next) => { ... });
```

Then in the advanced case, it could be something like this:

```javascript
app.use('/slack', everyauth.authorize('slack', {
   onComplete: async (res, res, ctx) => {
      ctx.identity.tags['fusebit.userId'] = 'usr-123';
      ctx.identity.tags['subscriptionLevel'] = 'gold';
      await everyauth.saveIdentity(ctx);
      res.redirect('/whatever');
   } 
}));

app.get('/whatever', (req, res, next) => { ... });
```

Reply:
... and the onAuthorized is then cut.